jekyll/_posts/2020-06-04-ray-tracing-in-nim.md

```diff
@@ -0,0 +1,316 @@
+---
+title: "Ray tracing in Nim"
+author: Mamy Ratsimbazafy (mratsim)
+excerpt: "I am convinced that if you want to start a rendering project from scratch, Nim is the best language to use."
+---
+
+<div class="sidebarblock">
+  <div class="content">
+    <div class="title">Guest post</div>
+    <div class="paragraph">
+      This is a guest post by Mamy Ratsimbazafy (mratsim). If you would like to publish articles as a guest author on nim-lang.org then get in touch with us via
+      <a href="https://twitter.com/nim_lang">Twitter</a> or <a href="https://nim-lang.org/community.html">otherwise</a>.
+    </div>
+  </div>
+</div>
+
+
+
+## Trace of Radiance
+
+Trace of Radiance is an implementation
+of [Ray Tracing in One Weekend](https://github.com/RayTracing/raytracing.github.io)
+in the Nim programming language.
+
+<p style="text-align: center;">
+  <img width="256" height="144" src="{{ site.baseurl }}/assets/news/images/mratsim-raytracing/book1_animation.gif">
+</p>
+
+The goals are:
+- Learn more about raytracing
+- Have fun
+- Serve as a testbed for own multithreading runtime, [Weave](https://github.com/mratsim/weave)
+- Showcase Nim capabilities
+
+In particular, I am convinced (and obviously biased) that if you want to start
+a rendering project (raytracing or otherwise) from scratch, Nim is the best
+language to use, in particular if you are focused on:
+- speed
+- correctness
+- compilation times and compile-time compute
+- approachability & development agility
+
+
+## Speed
+
+Rendering is an extremely time-consuming task.
+Nim is fast, very fast, even faster than C++ from time to time.
+
+Here are two benchmarks:
+
+
+### 1. Raytracing in one Weekend
+
+From the "RayTracing in One Weekend" first book, using:
+
+- Nim 1.2 (GCC 10.1), flag `-d:danger`
+- C++ with GCC 10.1, flag `-O3`
+
+On Intel Skylake-X i9-9980XE, overclocked at 4.1GHz all-core turbo.
+Compiled in x86-64 mode (SSE2 only), 384x216 image, 100 rays per pixel:
+
+| Nim     | C++     |
+| ------- | ------- |
+| 38.577s | 42.303s |
+
+The Nim sources for this benchmark can be retrieved from the
+[first release](https://github.com/mratsim/trace-of-radiance/tree/v0.1.0).
+
```

Comment:
It would be best to add the commands here

```Nim
git clone https://github.com/mratsim/trace-of-radiance
cd trace-of-radiance
git checkout v0.1.0
nim c -d:danger --outdir:build trace_of_radiance.nim
./build/trace_of_radiance > image.ppm
```